### PR TITLE
Wait for profile peek task to stop before exit

### PR DESCRIPTION
**Pull request by IanButterworth**
Trying to help with spurious reports of https://github.com/JuliaLang/julia/issues/50873

### CODE REVIEW

**Review by timholy — APPROVED** [review]
Closes #51003

**Comment by IanButterworth** [comment]
Before this precompiling DifferentialEquations I got
```
┌ ArrayInterface → ArrayInterfaceBandedMatricesExt
│  [pid 16932] waiting for IO to finish:
│   TYPE[FD/PID]       @UV_HANDLE_T->DATA
│   timer              @0x600002e75100->0x111324580
└  
┌ ExponentialUtilities
│  [pid 16870] waiting for IO to finish:
│   TYPE[FD/PID]       @UV_HANDLE_T->DATA
│   timer              @0x60000787af00->0x17c0dacb0
└  
┌ ChainRulesCore
│  [pid 16721] waiting for IO to finish:
│   TYPE[FD/PID]       @UV_HANDLE_T->DATA
│   timer              @0x6000003efb00->0x10dff95d0
└  
```
With this
```
┌ ArrayInterface → ArrayInterfaceBandedMatricesExt
│  [pid 63460] waiting for IO to finish:
│   TYPE[FD/PID]       @UV_HANDLE_T->DATA
│   timer              @0x600003efba00->0x110d5e560
└  
```

File: base/Base.jl [hunk 1]
```diff
@@ -567,6 +567,7 @@ if is_primary_base_module
 # triggers printing the report and (optionally) saving a heap snapshot after a SIGINFO/SIGUSR1 profile request
 # Needs to be in Base because Profile is no longer loaded on boot
 const PROFILE_PRINT_COND = Ref{Base.AsyncCondition}()
+profile_peek_task::Union{Nothing,Task} = nothing
```
Comment (vchuravy) [inline base/Base.jl:570]:
```suggestion
global profile_peek_task::Task
```

**Comment by vtjnash** [comment]
I don't get it. Where do you see a `Timer` or `sleep` call here?

**Comment by timholy** [comment]
You know more about this than I do, but just in case: in `Base.__init__` we spawn `profile_printing_listener` which waits on `PROFILE_PRINT_COND`. I don't fully follow the `Base.uv_unref(cond.handle)` operations, but presumably the notification doesn't always arrive before Julia tries to exit. On slack it was found that these warnings get printed disproportionately for fast-to-precompile "packages," which fits many extensions.

**Comment by vtjnash** [comment]
We can observe this does not seem to be the fix by generating an empty package:
```
(@v1.11) pkg> generate Hello

(@v1.11) pkg> dev ./Hello

julia> while true; Base.compilecache(Base.identify_package(Main, "Hello")); end
[ Info: Precompiling Hello [b78c8773-8b09-44b1-ab9b-12867abe6258]
("/home/vtjnash/.julia/compiled/v1.11/Hello/PkeV8_ikVpV.ji", "/home/vtjnash/.julia/compiled/v1.11/Hello/PkeV8_ikVpV.so")
```

**Comment by PallHaraldsson** [comment]
I don't like this for many reasons.

Such as it may not work. I found exit_on_sigint, and Julia exit is not always called to exit, rather the C function, and also that C function is exported, and if called should it also do this logic?

First a question. When your main strict ends, is exit() or similar called? I couldn't locate, and think you would want it called in almost all situations when you exit.

But I note C and C++ have many ways to exit a program, I would like to know how many Julia has, all the ways; e.g. similar to   _Exit() and std::quick_exit()?

https://en.cppreference.com/w/cpp/utility/program/quick_exit

https://stackoverflow.com/questions/9758495/what-is-the-difference-between-stdquick-exit-and-stdabort-and-why-was-stdq


I guess you need and want to have similar exit to C/C++'s default, I thought quick_exit would be faster/fastest, it seems not. I would like a very fast or immediate exit, and I fear that Tasks will take a long time to finish, or will not, deadlock. I suppose you still may not want to just abort by default but I would at least want to have that option.

**Comment by IanButterworth** [comment]
Sorry, I should've closed this. It doesn't seem to be the issue. The issue appears to be `Pkg.precompile` not getting the ordering of deps quite right, and the internal precompile process of packages hitting the pid locks when they shouldn't

**Pull request closed without merging**